tempodb/encoding/common/bloom_test.go

```diff
@@ -52,9 +56,60 @@ func TestShardedBloom(t *testing.T) {
 		if !found {
 			missingCount++
 		}
-		assert.Equal(t, found, filters[ShardKeyForTraceID(traceID)].Test(traceID))
+		assert.Equal(t, found, filters[ShardKeyForTraceID(traceID, b.GetShardCount())].Test(traceID))
 	}
 
 	// check that missingCount is less than bloomFP
 	assert.LessOrEqual(t, float64(missingCount), bloomFP*numTraces)
 }
+
+func TestShardedBloomFalsePositive(t *testing.T) {
+	tests := []struct {
+		name             string
+		bloomFP          float64
+		shardSize        uint
+		estimatedObjects uint
+	}{
+		{
+			name:             "regular",
+			bloomFP:          0.01,
+			shardSize:        100,
+			estimatedObjects: 1000,
+		},
+		{
+			name:             "large estimated objects",
+			bloomFP:          0.01,
+			shardSize:        100,
+			estimatedObjects: 10000,
+		},
+		{
+			name:             "large shard size",
+			bloomFP:          0.01,
+			shardSize:        100000,
+			estimatedObjects: 10,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			b := NewBloom(tt.bloomFP, tt.shardSize, tt.estimatedObjects)
+
+			// get byte representation
+			bloomBytes, err := b.Write()
+			assert.NoError(t, err)
+
+			// parse byte representation into willf_bloom.Bloomfilter
+			var filters []*willf_bloom.BloomFilter
+			for i := 0; i < b.GetShardCount(); i++ {
+				filters = append(filters, &willf_bloom.BloomFilter{})
+			}
+
+			for i, singleBloom := range bloomBytes {
+				_, err = filters[i].ReadFrom(bytes.NewReader(singleBloom))
+				assert.NoError(t, err)
+				assert.LessOrEqual(t, filters[i].EstimateFalsePositiveRate(tt.estimatedObjects/uint(b.GetShardCount())), tt.bloomFP)
```

Comment:
Added as a separate test because `EstimateFalsePositiveRate` wipes the bloom entries